// posthog-js/lib/src/session-props.js
/* Client-side session parameters. These are primarily used by web analytics,
 * which relies on these for session analytics without the plugin server being
 * available for the person level set-once properties. Obviously not consistent
 * between client-side events and server-side events but this is acceptable
 * as web analytics only uses client-side.
 *
 * These have the same lifespan as a session_id
 */
import { Info } from './utils/event-utils';
import { CLIENT_SESSION_PROPS } from './constants';
var generateSessionSourceParams = function (posthog) {
    return Info.personInfo({
        maskPersonalDataProperties: posthog === null || posthog === void 0 ? void 0 : posthog.config.mask_personal_data_properties,
        customPersonalDataProperties: posthog === null || posthog === void 0 ? void 0 : posthog.config.custom_personal_data_properties,
    });
};
var SessionPropsManager = /** @class */ (function () {
    function SessionPropsManager(instance, sessionIdManager, persistence, sessionSourceParamGenerator) {
        var _this = this;
        this._onSessionIdCallback = function (sessionId) {
            var _a;
            var stored = _this._getStored();
            if (stored && stored.sessionId === sessionId) {
                return;
            }
            var newProps = {
                sessionId: sessionId,
                props: _this._sessionSourceParamGenerator(_this.instance),
            };
            _this._persistence.register((_a = {}, _a[CLIENT_SESSION_PROPS] = newProps, _a));
        };
        this.instance = instance;
        this._sessionIdManager = sessionIdManager;
        this._persistence = persistence;
        this._sessionSourceParamGenerator = sessionSourceParamGenerator || generateSessionSourceParams;
        this._sessionIdManager.onSessionId(this._onSessionIdCallback);
    }
    SessionPropsManager.prototype._getStored = function () {
        return this._persistence.props[CLIENT_SESSION_PROPS];
    };
    SessionPropsManager.prototype.getSetOnceInitialSessionPropsProps = function () {
        var _a;
        var p = (_a = this._getStored()) === null || _a === void 0 ? void 0 : _a.props;
        if (!p) {
            return {};
        }
        if ('r' in p) {
            return Info.personPropsFromInfo(p);
        }
        else {
            return {
                $referring_domain: p.referringDomain,
                $pathname: p.initialPathName,
                utm_source: p.utm_source,
                utm_campaign: p.utm_campaign,
                utm_medium: p.utm_medium,
                utm_content: p.utm_content,
                utm_term: p.utm_term,
            };
        }
    };
    return SessionPropsManager;
}());
export { SessionPropsManager };
